airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/parquet_parser.py

```diff
@@ -2,21 +2,130 @@
+        # The indices column is an array of integers that maps to the values column
+        if pa.types.is_dictionary(parquet_value.type):
+            return {
+                "indices": parquet_value.indices.tolist(),
+                "values": parquet_value.dictionary.tolist(),
+            }
+
+        # Convert duration to seconds, then convert to the appropriate unit
+        if pa.types.is_duration(parquet_value.type):
+            duration = parquet_value.as_py()
+            duration_seconds = duration.total_seconds()
+            if parquet_value.type.unit == "s":
+                return duration_seconds
+            elif parquet_value.type.unit == "ms":
+                return duration_seconds * 1000
+            elif parquet_value.type.unit == "us":
+                return duration_seconds * 1_000_000
+            elif parquet_value.type.unit == "ns":
+                return duration_seconds * 1_000_000_000 + duration.nanoseconds
+            else:
+                raise ValueError(f"Unknown duration unit: {parquet_value.type.unit}")
+        else:
+            return parquet_value.as_py()
+
+    @staticmethod
+    def parquet_type_to_schema_type(parquet_type: pa.DataType) -> Mapping[str, str]:
+        """
+        Convert a pyarrow data type to an Airbyte schema type.
+        Parquet data types are defined at https://arrow.apache.org/docs/python/api/datatypes.html
+        """
+
+        if pa.types.is_timestamp(parquet_type):
+            return {"type": "string", "format": "date-time"}
+        elif pa.types.is_date(parquet_type):
+            return {"type": "string", "format": "date"}
+        elif ParquetParser._is_string(parquet_type):
+            return {"type": "string"}
+        elif pa.types.is_boolean(parquet_type):
+            return {"type": "boolean"}
+        elif ParquetParser._is_integer(parquet_type):
+            return {"type": "integer"}
+        elif pa.types.is_floating(parquet_type):
+            return {"type": "number"}
+        elif ParquetParser._is_object(parquet_type):
+            return {"type": "object"}
+        elif ParquetParser._is_list(parquet_type):
+            return {"type": "array"}
+        else:
+            raise ValueError(f"Unsupported parquet type: {parquet_type}")
+
+    @staticmethod
+    def _is_binary(parquet_type: pa.DataType) -> bool:
+        return bool(pa.types.is_binary(parquet_type) or pa.types.is_large_binary(parquet_type))
+
+    @staticmethod
+    def _is_integer(parquet_type: pa.DataType) -> bool:
+        return bool(pa.types.is_integer(parquet_type) or pa.types.is_duration(parquet_type))
+
+    @staticmethod
+    def _is_string(parquet_type: pa.DataType) -> bool:
+        return bool(
+            pa.types.is_time(parquet_type)
+            or pa.types.is_string(parquet_type)
+            or pa.types.is_large_string(parquet_type)
+            or pa.types.is_decimal(parquet_type)  # Return as a string to ensure no precision is lost
```

Comment:
Looks like this is different than what we do with the [current S3 connector](https://github.com/airbytehq/airbyte/blob/a47fcf174b8a63e80dd445369e5fd7f74924c567/airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/parquet_parser.py#L26). I'm tempted to keep it consistent so we don't break the existing connector when we switch over, but also don't want to be short-sighted by failing to change something if it's an improvement, just for the sake of backwards-compatibility.

I'm leaning toward keeping it like this for now, but identifying the S3 connectors that are currently syncing parquet files with decimal types, so we know what the impact will be (this can be done as part of the ticket for adapting the new file-based CDK to handle existing S3 connector configs). Then we can either alter the behavior depending on whether the connector is using the old config or new (e.g. old configs will still cast to number), or have affected customers do a full refresh. Thoughts?

Reply:
I think the current implementation is problematic. Converting decimals to floats can lead to loss of precision. I don't have a good way of checking how frequently users use this field since the schema in the config only defines the Airbyte type used (number).

I can add an optional flag to let the user configure how these fields should be represented to avoid introducing a breaking change

Reply:
@clnoll I'll add the flag in a separate PR to avoid the need to re-review this PR.